llama-index-integrations/tools/llama-index-tools-cassandra/llama_index/tools/cassandra/cassandra_database_wrapper.py

```diff
@@ -0,0 +1,675 @@
+"""Apache Cassandra database wrapper."""
+from __future__ import annotations
+
+import re
+import traceback
+from typing import Any, Dict, List, Optional, Sequence, Tuple, Union
+
+from cassandra.cluster import ResultSet, Session
+from pydantic import BaseModel, Field, root_validator
+
+
+IGNORED_KEYSPACES = [
+    "system",
+    "system_auth",
+    "system_distributed",
+    "system_schema",
+    "system_traces",
+    "system_views",
+    "datastax_sla",
+    "data_endpoint_auth",
+]
+
+
+class CassandraDatabase:
+    """Apache Cassandra® database wrapper."""
+
+    def __init__(
+        self,
+        session: Optional[Session] = None,
+        exclude_tables: Optional[List[str]] = None,
+        include_tables: Optional[List[str]] = None,
+        cassio_init_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        self._session = self._resolve_session(session, cassio_init_kwargs)
+        if not self._session:
+            raise ValueError("Session not provided and cannot be resolved")
+
+        self._exclude_keyspaces = IGNORED_KEYSPACES
+        self._exclude_tables = exclude_tables or []
+        self._include_tables = include_tables or []
+
+    def run(
+        self,
+        query: str,
+        fetch: str = "all",
+        include_columns: bool = False,
+        **kwargs: Any,
+    ) -> Union[str, Sequence[Dict[str, Any]], ResultSet]:
+        """Execute a CQL query and return the results."""
+        clean_query = self._validate_cql(query, "SELECT")
+        result = self._session.execute(clean_query, **kwargs)
+        if fetch == "all":
+            return list(result)
+        elif fetch == "one":
+            return result.one()._asdict() if result else {}
+        elif fetch == "cursor":
+            return result
+        else:
+            raise ValueError("Fetch parameter must be either 'one', 'all', or 'cursor'")
+
+    def run_no_throw(
+        self,
+        query: str,
+        fetch: str = "all",
+        include_columns: bool = False,
+        **kwargs: Any,
+    ) -> Union[str, Sequence[Dict[str, Any]], ResultSet]:
+        """Execute a CQL query and return the results or an error message."""
+        try:
+            return self.run(query, fetch, include_columns, **kwargs)
+        except Exception as e:
+            """Format the error message"""
+            return f"Error: {e}\n{traceback.format_exc()}"
+
+    def get_keyspace_tables_str_no_throw(self, keyspace: str) -> str:
```

Comment:
I would let the calling function deal with exception handling and remove the `*_no_throw` methods.

Reply:
Removed